tests/host_tools/memory.py

```diff
@@ -137,12 +126,39 @@ def run(self):
 
             time.sleep(self.MEMORY_SAMPLE_TIMEOUT_S)
 
-    def is_in_guest_mem_region(self, address):
-        """Check if the address is inside the guest memory region."""
-        if self._guest_mem_start is None:
-            return False
-        guest_mem_end = self._guest_mem_start + self.guest_mem_mib
-        return self._guest_mem_start <= address < guest_mem_end
+    def update_guest_mem_regions(self, address, size_kib):
+        """
+        If the address is recognised as a guest memory region,
+        cache it and return True, otherwise return False.
+        """
+
+        # If x86_64 guest memory exceeds 3328M, it will be split
+        # in 2 regions: 3328M and the rest. We have 3 cases here
+        # to recognise a guest memory region:
+        #  - its size matches the guest memory exactly
+        #  - its size is 3328M
+        #  - its size is guest memory minus 3328M.
+        if size_kib in (
+            self.guest_mem_mib * 1024,
+            self.X86_MEMORY_GAP_START,
+            self.guest_mem_mib * 1024 - self.X86_MEMORY_GAP_START,
+        ):
+            if not self._guest_mem_start_1:
+                self._guest_mem_start_1 = address
+                return True
+            if not self._guest_mem_start_2:
+                self._guest_mem_start_2 = address
+                return True
+        return False
+
+    def is_in_guest_mem_regions(self, address):
+        """Check if the address is inside a guest memory region."""
+        for guest_mem_start in [self._guest_mem_start_1, self._guest_mem_start_2]:
+            if guest_mem_start is not None:
+                guest_mem_end = guest_mem_start + self.guest_mem_mib
+                if guest_mem_start <= address < guest_mem_end:
+                    return True
```

Comment:
super nit, so feel free to ignore. I tend to prefer the following:

```suggestion
            if guest_mem_start is None:
                continue
            guest_mem_end = guest_mem_start + self.guest_mem_mib
            if guest_mem_start <= address < guest_mem_end:
                return True
```

Reply:
rewrote that anyway